File: DS/C++_Programming/SinglyCL/SinglyCL.cpp
#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;

class SinglyCL
{
    private:
        PNODE first;
        PNODE last;
        int iCount;
    
    public:

        SinglyCL()
        {
            first = NULL;
            last = NULL;
            iCount = 0;
        }

        void InsertFirst(int no)
        {
            PNODE newn = NULL;

            newn = new NODE;

            newn -> data = no;
            newn -> next = NULL;

            if((first == NULL) && (last == NULL))
            {
                first = newn;
                last = newn;
            }
            else
            {
                newn -> next = first;
                first = newn;
            }

            last -> next = first;
            iCount++;
        }

        void InsertLast(int no)
        {
            PNODE newn = NULL;

            newn = new NODE;

            newn -> data = no;
            newn -> next = NULL;

            if((first == NULL) && (last == NULL))
            {
                first = newn;
                last = newn;
            }
            else
            {
                last -> next = newn;
                last = newn;
            }

            last -> next = first;
            iCount++;
        }

        void InsertAtPos(int no, int iPos)
        {
            int iCnt = 0;
            PNODE newn = NULL;
            PNODE temp = NULL;

            if((iPos < 1) || (iPos > iCount + 1))
            {
                cout<<"Invalid Postion.";
                return;
            }

            if(iPos == 1)
            {
                InsertFirst(no);
            }
            else if(iPos == iCount + 1)
            {
                InsertLast(no);
            }
            else
            {
                newn = new NODE;

                newn -> data = no;
                newn -> next = NULL;

                temp = first;

                for(iCnt = 1; iCnt < iPos - 1; iCnt++)
                {
                    temp = temp -> next;
                }

                newn -> next = temp -> next;
                temp -> next = newn;
            }

            iCount++;
        }

        void DeleteFirst()
        {
            if((first == NULL) && (last == NULL))
            {
                return;
            }
            else if(first == last)
            {
                delete first;
                first == NULL;
                last ==NULL;
            }
            else
            {
                first = first -> next;
                delete last -> next;
            }

            last -> next = first;
            iCount--;
        }

        void DeleteLast()
        {
            PNODE temp = NULL;

            if((first == NULL) && (last == NULL))
            {
                return;
            }
            else if(first == last)
            {
                delete first;
                first == NULL;
                last ==NULL;
            }
            else
            {
                temp = first;

                while(temp -> next -> next != first)
                {
                    temp = temp -> next;
                }

                delete(temp -> next);
                last = temp;
            }

            last -> next = first;
            iCount--;
        }

        void DeleteAtPos(int iPos)
        {
            int iCnt = 0;
            PNODE target = NULL;
            PNODE temp = NULL;

            if((iPos < 1) || (iPos > iCount))
            {
                cout<<"Invalid Postion.";
                return;
            }

            if(iPos == 1)
            {
                DeleteFirst();
            }
            else if(iPos == iCount)
            {
                DeleteLast();
            }
            else
            {
                temp = first;

                for(iCnt = 1; iCnt < iPos - 1; iCnt++)
                {
                    temp = temp -> next;
                }

                target = temp -> next;
                temp -> next = target -> next;
                delete target;
            }

            iCount--;
        }

        void Display()
        {
            PNODE temp = NULL;

            temp = first;

            cout<<"\n -> ";

            do
            {
                cout<<" | "<<temp -> data<< " | -> ";
                temp = temp -> next;
            }while(temp != last -> next);

            cout<<"\n";
        }

        int Count()
        {
            return iCount;
        }
};

int main()
{
    int iRet = 0;
    SinglyCL obj;

    obj.InsertFirst(51);
    obj.InsertFirst(21);
    obj.InsertFirst(11);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    obj.InsertLast(101);
    obj.InsertLast(111);
    obj.InsertLast(121);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    obj.InsertAtPos(75, 4);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    obj.DeleteAtPos(4);

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    obj.DeleteFirst();

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    obj.DeleteLast();

    obj.Display();
    iRet = obj.Count();
    cout<<"Number of elements are : "<<iRet<< "\n";

    return 0;
}

File: DS/Java_Programming/SinglyCL/SinglyCL.java
class SinglyCLLNode
{
    public int data;
    public SinglyCLLNode next;

    public SinglyCLLNode(int no)
    {
        data = no;
        next = null;
    }
}

class SinglyCLL
{
    private SinglyCLLNode first;
    private SinglyCLLNode last;
    private int iCount;

    public SinglyCLL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {
        SinglyCLLNode newn = null;

        newn = new SinglyCLLNode(no);

        if((first == null) && (last == null))
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }

        last.next = first;

        iCount++;
    }

    public void InsertLast(int no)
    {
        SinglyCLLNode newn = null;

        newn = new SinglyCLLNode(no);

        if((first == null) && (last == null))
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            last = newn;
        }

        last.next = first;

        iCount++;
    }

    public void InsertAtPos(int no, int Pos)
    {
        SinglyCLLNode newn = null;
        SinglyCLLNode temp = null;
        int iCnt = 0;

        if((Pos < 1) || (Pos > iCount + 1))
        {
            System.out.println("Invalid position");
            return;
        }

        if(Pos == 1)
        {
            InsertFirst(no);
        }
        else if(Pos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            newn = new SinglyCLLNode(no);

            temp = first;

            for(iCnt = 1; iCnt < Pos - 1; iCnt++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if((first == null) && (last == null))
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
        }

        last.next = first;

        System.gc();

        iCount--;
    }

    public void DeleteLast()
    {
        SinglyCLLNode temp = null;

        if((first == null) && (last == null))
        {
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            temp = first;

            while(temp.next.next != first)
            {
                temp = temp.next;
            }

            last = temp;
        }

        last.next = first;

        System.gc();

        iCount--;
    }

    public void DeleteAtPos(int Pos)
    {
        SinglyCLLNode temp = null;
        int iCnt = 0;

        if( (Pos < 1) || (Pos > iCount) )
        {
            System.out.println("Invalid Position");
            return;
        }

        if(Pos == 1)
        {
            DeleteFirst();
        }
        else if(Pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            temp = first;

            for(iCnt = 1; iCnt < Pos - 1; iCnt++)
            {
                temp = temp . next;
            }

            temp.next = temp.next.next;

            System.gc();

            iCount--;
        }
    }

    public void Display()
    {
        SinglyCLLNode temp = null;

        temp = first;

        System.out.print("\n-> ");

        do
        {
            System.out.print("| " + temp.data + " | -> ");
            temp = temp.next;
        }while(temp != first);

        System.out.println();
    }

    public int Count()
    {
        return iCount;
    }

}
class SinglyCL
{
    public static void main(String A[])
    {
        SinglyCLL sobj = new SinglyCLL();
        int iRet = 0;

        sobj.InsertFirst(3);
        sobj.InsertFirst(2);
        sobj.InsertFirst(1);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.InsertLast(4);
        sobj.InsertLast(5);
        sobj.InsertLast(6);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.InsertAtPos(7, 4);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteAtPos(4);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteFirst();

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);

        sobj.DeleteLast();

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes are : " + iRet);
    }
}